File: #148_Sort_List/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

public class Solution {
    public static ListNode sortList(ListNode head) {
        if (head == null) return null;
        int len = 0;
        ListNode cur = head;
        while (cur != null) {
        	len++;
        	cur = cur.next;
        }
        if (len == 1) return head;
        int level = 0;
        int count = 1;
        while (count <= len) {
        	level++;
        	count *= 2;
        }
        ListNode dummyHead = new ListNode(0);
        dummyHead.next = head;
        for (int i = 0; i < level; i++) {
        	int jump = (int) Math.pow(2, i);
        	cur = dummyHead.next;
        	ListNode prev = dummyHead;
        	while (cur != null) {
        		ListNode list1 = cur;
        		for (int j = 0; j < jump; j++) {
        			if (j == jump - 1) {
        				ListNode tmp = cur;
        				cur = cur.next;
        				tmp.next = null;
        			}
        			else cur = cur.next;
        			if (cur == null) break;
        		}
    			if (cur == null) break;
        		ListNode list2 = cur;
        		for (int j = 0; j < jump; j++) {
        			if (j == jump - 1) {
        				ListNode tmp = cur;
        				cur = cur.next;
        				tmp.next = null;
        			}
        			else cur = cur.next;
        			if (cur == null) break;
        		}
        		ListNode newHead = mergeTwoSortedList(list1, list2);
        		prev.next = newHead;
        		ListNode last = findLast(newHead);
        		prev = last;
        		last.next = cur;
        	}
        }
        return dummyHead.next;
    }

    public static ListNode mergeTwoSortedList(ListNode node1, ListNode node2) {
    	ListNode dummyHead = new ListNode(-1);
    	if (node1 == null) return node2;
    	if (node2 == null) return node1;
    	ListNode cur1 = node1;
    	ListNode cur2 = node2;
    	ListNode cur = null;
    	while (cur1 != null && cur2 != null) {
    		ListNode smaller = (cur1.val < cur2.val) ? cur1 : cur2;
    		if (dummyHead.next == null) {
    			dummyHead.next = smaller;
    			cur = smaller;
    		}
    		else {
    			cur.next = smaller;
    			cur = cur.next;
    		}
    		if (cur1 == smaller) cur1 = cur1.next;
    		else cur2 = cur2.next;
    	}
    	ListNode res = (cur1 != null) ? cur1 : cur2;
    	while (res != null) {
    		cur.next = res;
    		cur = cur.next;
    		res = res.next;
    	}
    	return dummyHead.next;
    }

    private static ListNode findLast(ListNode head) {
    	if (head == null) return null;
    	ListNode cur = head;
    	while (cur.next != null) {
    		cur =cur.next;
    	}
    	return cur;
    }

    public static void main(String[] args) {
    	ListNode node0 = new ListNode(5);
    	ListNode node1 = new ListNode(1);
    	// ListNode node2 = new ListNode(7);
    	// ListNode node3 = new ListNode(2);
    	// ListNode node4 = new ListNode(8);
    	// ListNode node5 = new ListNode(10);
    	// ListNode node6 = new ListNode(3);
    	// ListNode node7 = new ListNode(4);
    	node0.next = node1;
    	// node1.next = node2;
    	// node2.next = node3;
    	// node3.next = node4;
    	// node4.next = node5;
    	// node5.next = node6;
    	// node6.next = node7;
    	ListNode result = Solution.sortList(node0);
    	ListNode cur = result;
    	while (cur != null) {
    		System.out.print(cur.val + " ");
    		cur = cur.next;
    	}
    	System.out.println();
    }
}

File: #148_Sort_List/Solution.py
# Definition for singly-linked list.
class ListNode:
	def __init__(self, x):
		self.val = x
		self.next = None

class Solution:
	# @param {ListNode} head
	# @return {ListNode}
	def sortList(self, head):
		dummy = ListNode(0)
		dummy.next = head
		sz = 1               # size of the list that will be merged
		length = self.lenList(head)
		while sz < length:   # if the whole list is not long enough for the other sub list, merge is over
			cur = dummy.next # start from first node
			last = dummy     # tail node of previous merged lists
			while cur:
				prev = last
				h1 = cur     # first sub list
				for x in range(sz):
					if not cur:
						break
					cur = cur.next
					prev = prev.next
				tmp = prev   # end of first sub list
				h2 = cur     # second sub list
				if not h2:   # no second sub list, break
					break
				for x in range(sz):
					if not cur:
						break
					cur = cur.next
					prev = prev.next
				tmp.next = None  # cut out first sub list
				prev.next = None # cut out second sub list
				[newH, newT] = self.mergeTwoList(h1, h2)
				last.next = newH
				newT.next = cur
				last = newT  # update tail node of previously merged lists
			sz <<= 1
		return dummy.next


	# merge two sorted list, return [head, tail]
	def mergeTwoList(self, h1, h2):
		if not h1:
			return h2
		if not h2:
			return h1
		dummy = ListNode(0)
		c1, c2 = h1, h2
		cur = dummy
		while c1 or c2:
			if c1 and c2:
				if c1.val < c2.val:
					cur.next = c1
					c1 = c1.next
				else:
					cur.next = c2
					c2 = c2.next
			elif c1:
				cur.next = c1
				c1 = c1.next
			else:
				cur.next = c2
				c2 = c2.next
			cur = cur.next
		return [dummy.next, cur]

	def lenList(self, head):
		count = 0
		while head:
			count += 1
			head = head.next
		return count

	def traverse(self, head):
		while head:
			print head.val,
			head = head.next

sol = Solution()
n0 = ListNode(3)
n1 = ListNode(4)
n2 = ListNode(1)
n3 = ListNode(5)
n0.next = n1
n1.next = n2
n2.next = n3
sol.traverse(sol.sortList(n0))

# n0 = ListNode(1)
# n1 = ListNode(4)
# n2 = ListNode(2)
# n3 = ListNode(3)
# n0.next = n1
# n2.next = n3
# sol.traverse(sol.mergeTwoList(n0, n2)[0])